executor/write.go

```diff
@@ -65,50 +70,50 @@ func updateRecord(ctx context.Context, h int64, oldData, newData []types.Datum,
 			}
 			t.RebaseAutoID(val, true)
 		}
-		casted, err := table.CastValue(ctx, newData[i], col.ToInfo())
+		cmp, err := newData[i].CompareDatum(sc, oldData[i])
```

Comment:
Suppose the column `v` is `INTEGER` and we run `UPDATE t SET v = 1.0`, the new value `1.0` with type `DECIMAL` has already cast to `1 ` with type `INTEGER` before here , at line **57**. So it will be not changed.